File: LeetCodeContests/139/kotlinSolutions/src/gcdOfStrings.kt
class Solution
{
    fun gcd(a: Int, b: Int): Int = if (b == 0) a else gcd(b, a % b)

    fun gcdOfStrings(str1:String, str2:String):String
    {
        var lens = gcd(str1.length, str2.length)
        var str3 = str1.substring(0, lens)
        if((isMatch(str1, str3)) and isMatch(str2, str3)) return str3
        return ""
    }


    fun isMatch(s1:String, s2:String):Boolean
    {
        for(i in s1.indices)
        {
            if (s1[i] != s2[i % s2.length]) return false
        }
        return true
    }


}

fun main(args:Array<String>)
{
    println(Solution().gcdOfStrings("ABCABC", str2 = "ABC"))
}

File: LeetCodeContests/139/gcd_of_strings.py
from math import gcd
class Solution():
    def is_match(self, str1, str2):
        for i in range(len(str1)):
            if str1[i] != str2[i % len(str2)]: return False
        return True

    def gcdOfStrings( self, str1, str2):
        len1 = len(str1)
        len2 = len(str2)
        lene = gcd(len1, len2)
        str3 = str1[:lene]
        if self.is_match(str1, str3) and self.is_match(str2, str3): return str3
        return ""



